add 27255 notes (#652)
bitcoin-core-review-club/website: _posts/2023-04-05-#27255.md

commit:
---

## Notes

### Background and prior work

_Miniscript_ is a language for writing (a subset of) Bitcoin Scripts in a structured way, enabling analysis, composition, generic signing and more. _Miniscript_ support in Bitcoin Core has been an ongoing effort, with watch-only support for miniscript in P2WSH descriptors added in `v24.0.1`. For background:

* [https://bitcoin.sipa.be/miniscript](https://bitcoin.sipa.be/miniscript) gives an overview of the miniscript langauge
* [#24147](https://github.com/bitcoin/bitcoin/pull/24147) provides the rationale and initiale backbone needed for miniscript support in Bitcoin Core
* [#24148](https://github.com/bitcoin/bitcoin/pull/24148) adds watch-only support for miniscript descriptors
* [#24149](https://github.com/bitcoin/bitcoin/pull/24149) adds signing support for miniscript descriptors

> ### TapMiniscript

Author: slight preference for TapMiniscript

> ### TapMiniscript

Reviewer: okay then i'll leave as is to mirror the current PR title

commit:
---

## Notes

### Background and prior work

> _Miniscript_ is a language for writing (a subset of) Bitcoin Scripts in a structured way, enabling analysis, composition, generic signing and more. _Miniscript_ support in Bitcoin Core has been an ongoing effort, with watch-only support for miniscript descriptors added in `v24.0.1`. For background:

Reviewer 2: Maybe note that Miniscript descriptors are only supported under P2WSH in Bitcoin Core and Miniscript itself was anyway only really defined for P2WSH context, not any Bitcoin Script context.

Author: I mentioned P2WSH, but I don't want to confuse anyone by trying to explain it further

* [#24148](https://github.com/bitcoin/bitcoin/pull/24148) adds watch-only support for miniscript descriptors
* [#24149](https://github.com/bitcoin/bitcoin/pull/24149) adds signing support for miniscript descriptors

### TapMiniscript

While [_Tapscript_](https://github.com/bitcoin/bips/blob/master/bip-0342.mediawiki) shares most operations with legacy Bitcoin Script, there are a few notable differences. As such, _TapMiniscript_ has been a project to support tapscript in miniscript. For background:

* [This gist](https://gist.github.com/sipa/06c5c844df155d4e5044c2c8cac9c05e) contains the initial design and discussion on supporting miniscript in tapscript
* [#134](https://github.com/sipa/miniscript/pull/134/) adds tapscript support in miniscript

### Putting it all together

Putting it all together, this PR adds support for Tapscript in Miniscript in Bitcoin Core and makes Miniscript available inside `tr()` descriptors, both with watching and signing support.

## Questions

1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?

> 1. In your own words, what are some of the differences between Segwit v0 Script (Bitcoin Script with the Segwit Soft Fork rules) and Segwit v1 script (Tapscript)?

Reviewer 2: Maybe let's ask the question in the context of Miniscript?

```suggestion
1. In your own words, what are some of the differences between Segwit v0 Script (Bitcoin Script with the Segwit Soft Fork rules) and Tapscript rules that may be relevant to Miniscript? Why?
```

Also nit i removed Segwit v1 because Tapscript is only defined for a specific leaf, in the future another Segwit v1 output with another leaf version used may have different Script semantics.

> 1. What is Miniscript?

Reviewer 2: But it's already defined in the introduction above? Is it really necessary to ask this?

Also meta comment, here and below, why aren't the indices of the bullet points increasing?

Author: in markdown, bullet numbers dont matter. the renderer adds the numbers

Author: also good point re: miniscript, will remove.

> 1. What is Tapscript?

Reviewer 2: Good point to ask this maybe a good opportunity to clear up any confusion between leaf version, segwit version, etc..

> 1. This PR [adds](https://github.com/bitcoin/bitcoin/pull/27255/commits/c0ba8ebbf6369b37b645165bb5cd638fc7eee67f) a `multi_a` fragment for tapscript. How is this different than the existing `multi` fragment? Does it have the same properties as `multi`?

Reviewer 2: Nit
```suggestion
1. This PR [adds](https://github.com/bitcoin/bitcoin/pull/27255/commits/c0ba8ebbf6369b37b645165bb5cd638fc7eee67f) a `multi_a` fragment for tapscript. How is this different than the existing `multi` fragment? Does it have the same properties as `multi`? Why?
```

1. In miniscript, we have type modifiers, which guarantee addititional properties for an expression. [866284d](https://github.com/bitcoin/bitcoin/pull/27255/commits/866284d007993551f681809d9e48175a3b0fe0c1) makes "**d**" have the "**u**" property under Tapscript.
	* What are the "**d**" and "**u**" type modifiers?
	* Why is it that we can make **d** have the **u** property here? Why not in non-tapscript miniscript?

> 1. The majority of complexity in this PR comes from statically ensuring no spending path exceeds the stack size at execution time:

Reviewer 2: nit: i know it's my words but could we get rid of "complexity"? I'm not sure it's really that complex, and there are other things that may seem complicated, idk.
```suggestion
1. This PR adds some logic for statically ensuring no spending path exceeds the stack size at execution time:
```

	* Why does this matter for Tapscript?
	* What's the approach taken by this PR? What are the pros/cons? (hint: [efdd154](https://github.com/bitcoin/bitcoin/pull/27255/commits/efdd1543597aff49c56a1abaa75b574be3b330db))
	* Can you think of an alternative approach to ensure no spending path exceeds the stack size?

> 1. In [e81635c](https://github.com/bitcoin/bitcoin/pull/27255/commits/e81635c39d99a158629544fefd765b3994f3d7c4), the scripts are optionally padded during fuzzing. Why is this necessary?

Reviewer 2: ```suggestion
1. In [e81635c](https://github.com/bitcoin/bitcoin/pull/27255/commits/e81635c39d99a158629544fefd765b3994f3d7c4), the scripts are optionally padded during fuzzing. Why?
```

It's not necessary, it's just more efficient
